samples/bluetooth/eddystone/src/main.c

```diff
@@ -490,7 +490,7 @@ static ssize_t write_adv_data(struct bt_conn *conn,
 		 */
 		slot->ad[2].data_len = MIN(slot->ad[2].data_len,
 					   len + EDS_URL_WRITE_OFFSET);
-		memcpy(&slot->ad[2].data + EDS_URL_WRITE_OFFSET, buf,
+		memcpy((uint8_t *) slot->ad[2].data + EDS_URL_WRITE_OFFSET, buf,
```

Comment:
this is quite questionable to begin with ... `data` is `const uint8_t *`.

Reply:
> this is quite questionable to begin with ... `data` is `const uint8_t *`.

Yeah, someone who understands what this is supposed to be doing would be a big help. All I've fixed is to replace an obvious memory smashing operation with something that might end up trying to write to read-only memory?

Reply:
I believe @keith-packard fix is correct and technically sound. The object that is being written to is non-const, so this cast is allowed.

I agree that a cast like this is horrible and needs a refactor. But I am for adding this quick fix. The original code is obviously wrong.

Reply:
I did a bit of studying of this sample app, and the `slot->ad` array is defined as follows:
```c
static struct eds_slot eds_slots[NUMBER_OF_SLOTS] = {
	[0 ... (NUMBER_OF_SLOTS - 1)] = {
		...
		.ad = {
			BT_DATA_BYTES(BT_DATA_FLAGS, BT_LE_AD_NO_BREDR),
			BT_DATA_BYTES(BT_DATA_UUID16_ALL, 0xaa, 0xfe),
			BT_DATA_BYTES(BT_DATA_SVC_DATA16,
				      0xaa, 0xfe, /* Eddystone UUID */
				      0x10, /* Eddystone-URL frame type */
				      0x00, /* Calibrated Tx power at 0m */
				      0x00, /* URL Scheme Prefix http://www. */
				      'z', 'e', 'p', 'h', 'y', 'r',
				      'p', 'r', 'o', 'j', 'e', 'c', 't',
				      0x08) /* .org */
		},
```
It's defined as part of parent struct that's declared static, but not const, so I don't think there's read-only memory involved.

The `bt_data` struct and the `BT_DATA_BYTES()` macro look as follows:
```c
struct bt_data {
	uint8_t type;
	uint8_t data_len;
	const uint8_t *data;
};

#define BT_DATA(_type, _data, _data_len) \
	{ \
		.type = (_type), \
		.data_len = (_data_len), \
		.data = (const uint8_t *)(_data), \
	}

#define BT_DATA_BYTES(_type, _bytes...) \
	BT_DATA(_type, ((uint8_t []) { _bytes }), \
		sizeof((uint8_t []) { _bytes }))
```

So the `slot->ad[2].data` pointer is indeed indicating read-only memory through its type, even though the memory behind it is not actually read-only in the case of this sample app. So to me your change does look like it should result in working code.

That said, casting from `const uint8_t *` to `uint8_t *` always looks a bit dangerous, so the sample might be better implemented by *not* using the `BT_DATA_BYTES()` macro, rather the `uint8_t` array for the data should have been separately declared as a named writable array elsewhere and then given directly to the `BT_DATA()` macro, and then that new variable should be used to access it (for writing) rather than getting to it through the bt_data.data pointer, which implies read-only memory. Anyway, that's beyond the scope of this PR.